### ISSUE TITLE
Lua serialise improvements

### ISSUE THREAD

**Issue by aidanholm** [luakit/luakit#377]
- [ ] deduplicate serialised tables
- [ ] serialise metatables as well
- [ ] add upvalues to serialised functions
- [ ] on error, show path to bad value, e.g. `foo.bar[25].baz`

**Comment by szbnwer** [comment]
deduplicating means to not clone a table with two refsin the same structure? or not die on a recursion?

how can functions and their upvalues serialized? this is something i wanna know for a long while :D but putting a c function anywhere into a structure can kill anything, and upvalues can be modified from more than one direction, so this can also mess a lot... my best idea about these stuffs was to not use anything tricky around stuffs i'd ever like to serialize in the future :D and my best idea to serialize a function is to get it's location in the sauce, so i have it as text, or sumtin like myFunction=regFunc'function(ftw) return 1048576 end' ((no, i don't like perl)) where regFunc could dostring the code and make it a key to a table where the value is the original string and it's implementation is childgame, serializing with it just as well much straightforward...

however i still didn't implement this, because i'm worrying, that it would just mess up my life without real usefulness :D i think tables numbers strings and booleans can be enough to save any info, but most likely not for sending, because of functions, but still much easier to send actual codes in this case :D if i'd make an anon function hidden in a closure that i think it could be useful outside, then i'd modify my codes... - so what's the actul usecase?

and i'm keepin away myself from that common fetish so called metatables because i don't wanna mess up my data...

currently in my app, Szoszi (which is still private), i wanna make something like project managing to bring all my works to the same ground, where i wanna save states and options and have custom bindings and handlers, so now i'm also facing with something similar.. because i wanna add a new project like a function call with parameters like basedir for get all of it's files, a name for the whole and custom handlers and prebuilt handlers, where custom handlers will be tricky this way, probably the best i can have is to add them to the table where the handlers of a file/type/project will be and if everything works fine and they are reusable for other projects then i can move the codes by hand, but this is just still not the best as i wanna see some kinda automagic everywhere :D probably for this usecase that regFunc stuff could be fine...

anyway i've seen debug stuffz that are possibly usable here, but is there any better way to go, or it depends, or whats your idea and goal?

**Comment by szbnwer** [comment]
anyway i can offer you this one:

```lua
function(path) -- array table
  if type(path)=='table' then
    for i, p in ipairs(path) do
      if type(p)=='string' and p:match'^[_%a]+[_%w]*$' then
        if i~=1 then out'.' end
        out(p)
      else
        out'['
        if type(p)=='string' then
          out("'", p, "'")
        elseif type(p)=='number' or type(p)=='boolean' then
          out(tostring(p))
        else
          out(tostring(p)) end
        out']' end end end end
```

the out can be replaced by whatever you want, i have an output buffer behind that and will be a function that will vary as it will serve as console output and tekui graphical output and server output as well :D yup i like mad stuffs :D the elseif block can be deleted as unnecessary, just those are straightforward and everything that remains won't work like foo.bar[25].baz out of the box... so i've thought about how could i handle those and this stayed unoptimized... and you can give it a name for it as well, cuz mine was simply path because it's place gave complete sense for this simple name...

i only dunno that if you will use it under gpl then i can have the same under different licence in my project?

**Comment by aidanholm** [comment]
When serializing Lua data for IPC, we currently follow the reference graph in a depth-first manner, without remembering where we've been. This causes infinite recursion if the reference graph is cyclic, and causes item duplication if the reference graph is not a tree but a DAG:
```lua
local A = {}
local B = {A, A}
assert(B[1] == B[2]) -- True, passes
... --- Serialize and deserialize....
assert(B[1] == B[2]) -- False, error! B[1] and B[2] are now two different copies of the original, not one copy
```

Functions and serialized using the code in `common/luaserialize.c`; upvalues aren't yet, but they're just regular Lua data so it'll be the same. Lua provides some C APIs that make this fairly simple.

The serializer will only work on Lua data (otherwise an error is raised), and serialization is for message passing only; there is no keeping objects in sync across processes. When you deserialize, you get a copy of the original.

For 90% of uses, tables, numbers, strings, and booleans are enough; but this API is provided by luakit, so it has to cover the other 10% as well. It'll be very useful for the formfiller API, in order to simplify the interface: users don't have to know that their functions are being sent between processes, it should just work.

Metatables are fun :) Very powerful construct; good for creating read-only tables for example. Although, there is a bit of unseen magic about them.

That code is neat! Everything for this is written in C, so I'll implement the "path to bad value" by unwinding the call stack on error and pushing the Lua key from each stack frame onto an array.

As far as licensing goes, if you own the copyright on that code you can grant me a GPLv3-compatible license to use that code, while separately licensing that code in Szoszi or other projects however you like. But I wouldn't worry too much about it, as I'll have to write the equivalent in C, and any code like that is going to end up looking pretty similar, because there's only really one sensible way to write it.

**Comment by szbnwer** [comment]
yep you can use my code as you like it, i'm only worriing about my project as a whole but this is the 3rd small chunk i'm offering for any purpose, i have a structure handler what can handle printing resoultz and handle lua tables and fs and whatever i wanna add it can iterate it can't die on recursions, nor repetions, it's not recursive, can be added any kinda limits filtering and whatever i want while it's running, and it will serve as search, diff, dump, serialize, edit by rules, deepcopy and kinda much anything :D this was sumtin i worked for so much months to be alive :D so these kinda evil monsters needed to held back to give value for my project when i will publicate it :D it's whole install is for a fresh raw debian, it's asking for a sudoer or root pw for installing git, make and the like needed, then the rest works without privileges, that is downloading sauces compiling and installing them into it's folder and running it... the app can be freely moved on fs without reinstall, it even works on the same architecture without install anything if it's installed once, it can be restarted, can lift it's privileges, and developed with live coding without loss from itself....... and much more :D but i have still 4k+ lines of plans to make it usable for all the man of mold (~mortals - so those who don't have time to learn all those horribly much stuffs that i've learnt, just as other programmers did...)

soooo about protection - you can get the parents and protect against recursion by checking equality for them, and for repetitive matery you can make a table with ref=true and for new elements you can check for existence in this table - this is fast, but not memory efficient, other way is to go through all the previous elements for match, but as i'm flatting first a whole structure, the 1st method have additional data, the 2nd not... but for a simple recursive walk through these are not this much easy tasks :dagger: 

anyway what c functions would be good for serialization? i'm reading much c, most of them are around lua world, but as i prefer luajit, i've never played with c api, because the lack of need...

and the other topic will get answer later, because i just got a lotsa todo.... :/
